### PR TITLE
Aligned with New CIP-30 Extension Scheme

### PR DESCRIPTION
**Pull request by Ryun1**
### Context
- Following the merger of #446 CIP-30 contains a scheme for extensions to its API.
- This means that CIP-62 can use the scheme as defined in CIP-30 rather than its own namespace (catalyst).

### Changes
- Removed initial API components:
  - `cardano.{walletName}.catalyst.apiVersion`
  - `cardano.{walletName}.catalyst.enable(purpose: VotingPurpose[])`
- Added:
  - `api.getVotingPurposes()`

- Prose:
  - Added description of connection flow to `Examples of Message Flows and Processes`.
  - Added description of missing voting flow to `Examples of Message Flows and Processes`.
  - Supporting prose throughout .

----------------------------------------------------------

[Rendered Version](https://github.com/Ryun1/CIPs/tree/cip-62-extension-rewrite/CIP-0062)

### CODE REVIEW

File: CIP-0062/README.md [hunk 1]
```diff
@@ -299,6 +265,18 @@ The API chosen here is for the minimum API necessary for dApp <-> Wallet
 interactions without convenience functions that don't strictly need the wallet's
 state to work.
 
+#### api.getVotingPurposes(): Promise\<VotingPurpose>[]
+
+Errors: [`APIError`](#extended-apierror)
+
+This is a chance for dApps to ask wallets what `VotingPurpose`s the user wishes to allow the dApp access to. It is up to wallet implementors if they deem it necessary to prompt the user for their permission to share supported Purposes.
```
Comment (Scitz0) [inline CIP-0062/README.md]: As this is just defined as an endpoint to fetch supported purposes by the wallet, there is no access request needed. 

Else endpoint would need to be redefined as `enableVotingPurposes(purpose: VotingPurpose[])` that takes a list of purposes to enable just like previous `catalyst.enable(purpose: VotingPurpose[])` call. 

But I dont see this as needed, I happy with wallet just returning a list of purposes it support. Its enough that the user ask for CIP62 permission once, when calling CIP30 enable(). Else it will result in a bad UX.

So text here should be updated to remove info about extra access request.

Comment (Ryun1) [reply CIP-0062/README.md]: @Scitz0 

I agree -- change made.

File: CIP-0062/README.md [hunk 2]
```diff
@@ -246,59 +247,36 @@ All TxSignErrors defined in [CIP-30](https://cips.cardano.org/cips/cip30/#txsign
...
-Upon successful connection via
-[`cardano.{walletName}.catalyst.enable()`](#cardanowalletnamecatalystenablepurpose-votingpurpose-promiseapi),
-a javascript object we will refer to as `API` (type) / `api` (instance) is
-returned to the dApp with the following methods.
+### Catalyst API Extension to CIP-30
 
-### Catalyst API
+The following endpoints extend [CIP-30's Full API](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0030#full-api).
 
 Except `signVotes`, no other method should require any user
```
Comment (refi93) [inline CIP-0062/README.md]: shouldn't also `submitDelegation` be mentioned here?

Comment (Ryun1) [reply CIP-0062/README.md]: Yup good spot - fixed.

File: CIP-0062/README.md [hunk 3]
```diff
@@ -246,59 +247,36 @@ All TxSignErrors defined in [CIP-30](https://cips.cardano.org/cips/cip30/#txsign
...
-[`api.signData()`](#apisubmitdelegationdelegation-delegation-promisesigneddelegationmetadata)
+[`api.submitDelegation()`](#apisubmitdelegationdelegation-delegation-promisesigneddelegationmetadata)
 must request the user's consent in an informative way for each and every API
 call in order to maintain security.
 
 The API chosen here is for the minimum API necessary for dApp <-> Wallet
 interactions without convenience functions that don't strictly need the wallet's
 state to work.
 
+#### api.getVotingPurposes(): Promise\<VotingPurpose>[]
```
Comment (refi93) [inline CIP-0062/README.md]: I guess this is just a typo
```suggestion
#### api.getVotingPurposes(): Promise\<VotingPurpose[]>
```

**Comment by refi93** [comment]
unrelated to this PR but I think there is an inconsistency in how `choice` is currently defined:

> * choice - This value MUST match one of the available `voteOptions` in the `proposal` element. An `UnknownChoiceError` should be thrown if the value is not one of the valid `voteOptions` of the `proposal`.

However this is the current definition of the `voteOptions` field in `Proposal`:

> voteOptions - Total number of vote options. Only present in a private/encrypted vote.

I guess the `choice` definition is based on an outdated definition of `voteOptions` when it was a list of numbers, but maybe I'm missing something. Can you update the docs to make this consistent?

File: CIP-0062/README.md [hunk 4]
```diff
@@ -368,19 +347,28 @@ The [Signed Delegation Metadata](#signeddelegationmetadata) of the voter registr
...
+2. **Connection** - Using the name of the selected wallet the dApp can call [CIP-30's enable](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0030#cardanowalletnameenable-extensions-extension----promiseapi) passing in this CIPs identifier: `cardano.{walletName}.enable([{"cip": 62 }])`. This should cause the wallet to ask the user to grant permission for the dApp connecting and using the CIP-62 API Extension.
+3. **Purpose** - Once connection is established the dApp will want to establish the supported wallet's `VotingPurpose`s, this is achieved via `.getVotingPurposes()`. By knowing the supported purposes the dApp can correctly choose which ones to use when calling `.signVotes()` and `.submitDelegation()`.
+
 #### Delegation
 
 Recall from [CIP-36](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0036) a registration is a self-delegation, allocating one's voting power to one's own voting key.
 
 1. **Get Voting Key** - dApp calls the method `api.getVotingCredentials()` to return the connected wallet account's public `voteKey`.
```
Comment (refi93) [inline CIP-0062/README.md:360]: not related to this PR, but I was just polishing HW wallet support when I realized that only Ledger HW wallet will be able to return the voting key. So this call would fail for a Trezor user even though they may want just to delegate their voting rights and for that usecase it's sufficient to get the staking key. One solution I can think of is to split this call into two, so each piece of information is available independently, i.e. `getVotingKey` and `getStakingKey` - what do you think @Ryun1 ?

Alternatively, we may make the `votingKey` in the returned value optional though that may not lead to good UX as it may no longer be clear why the votingKey wasn't returned as there would be no error code/message to indicate that

Comment (Ryun1) [reply CIP-0062/README.md:360]: @stevenj

Comment (refi93) [reply CIP-0062/README.md:360]: for the record, as we discussed that offline - it seems that the `stakingKeyPub` may not need to be exposed at all as the wallet can handle it internally. For the sake of constructing the delegation transaction, it's enough for the wallet to receive the staking address in the `DelegatedCertficate` [structure](https://github.com/ehanoc/CIPs/blob/39d575ca0f7604d7be2d4d5a9ab28d750b39006e/CIP-0062/README.md?plain=1#L151) and it can map it back to the respective public key - wallets need to do much heavier reverse mapping under the hood already anyway to support signing Cardano transactions with hardware wallets as hw wallets expect derivation paths, not raw addresses/keys.

If we suppose that a wallet has a single staking key, not even the staking address would be needed, but it seems to be more future proof and consistent with CIP-30 to include it as the CIP-30 API itself exposes the reward (i.e. staking) addresses as an array: https://github.com/cardano-foundation/CIPs/blob/master/CIP-0030/README.md?plain=1#L305

So to sum it up - unless there's some usecase I'm missing, I think the `stakingKeyPub` can be dropped from the `getVotingCredentials()` call altogether in which case it probably makes sense to rename the call back to `getVotingKey()` and the `submitDelegation()` call's parameter could reference just the `stakingAddress` which can be retrieved from the CIP-30 `getRewardAddresses()` call. This would also solve the problem raised in the comment above
